src/test/utils.test.ts

```diff
@@ -410,4 +415,145 @@ describe('Provable utilities - unit tests', () => {
+
+      const computedHash = Poseidon.hash([...outputNumbers, salt]);
+      const solutionHash = proof.publicOutput.solutionHash;
+
+      expect(separatedHistory).toEqual(attemptList);
+      expect(solutionHash).not.toEqual(computedHash);
+      expect(outputNumbers.map(Number)).toEqual([8, 3, 5, 2]);
+    });
+
+    it('Should generate a game where codeBreaker wins', async () => {
+      const round = 7;
+      const winnerFlag = 1;
+      const actions = guessConfig1;
+      const salt = codeMasterSalt;
+
+      const secretCombination = actions.secret;
+      const lastGameProof = await StepProgramCreateGame(
+        secretCombination,
+        salt,
+        codeMasterKey
+      );
+
+      const proof = await generateTestProofs(
+        round,
+        winnerFlag,
+        actions,
+        lastGameProof,
+        codeBreakerKey,
+        codeMasterKey,
+        salt
+      );
+
+      const publicOutputs = proof.publicOutput;
+
+      const outputNumbers = separateCombinationDigits(publicOutputs.lastGuess);
+      const history = deserializeCombinationHistory(
+        publicOutputs.packedGuessHistory
+      );
+      const separatedHistory = Array.from({ length: round }, (_, i) =>
+        separateCombinationDigits(history[i]).map(Number)
+      );
+
+      // Getting until round - 1, since in proof generation first round - 1 elements are used before secret combination.
+      const attemptList = actions.totalAttempts.slice(0, round - 1);
+      attemptList.push(secretCombination);
+
+      const computedHash = Poseidon.hash([...outputNumbers, salt]);
+      const solutionHash = publicOutputs.solutionHash;
+
+      expect(separatedHistory).toEqual(attemptList);
+      expect(solutionHash).toEqual(computedHash);
+      expect(outputNumbers.map(Number)).toEqual([6, 3, 8, 4]);
+      expect(BigInt(round)).toEqual(
+        publicOutputs.turnCount.sub(1).div(2).toBigInt()
+      );
+    });
+    it('Should generate a proof that is not solved yet.', async () => {
+      const round = 7;
+      const winnerFlag = 2;
+      const actions = guessConfig1;
+      const salt = codeMasterSalt;
+
+      const secretCombination = actions.secret;
+      const lastGameProof = await StepProgramCreateGame(
+        secretCombination,
+        salt,
+        codeMasterKey
+      );
+
+      const proof = await generateTestProofs(
+        round,
+        winnerFlag,
+        actions,
+        lastGameProof,
+        codeBreakerKey,
+        codeMasterKey,
+        salt
+      );
+
+      const publicOutputs = proof.publicOutput;
+
+      const outputNumbers = separateCombinationDigits(publicOutputs.lastGuess);
+
+      const computedHash = Poseidon.hash([...outputNumbers, salt]);
+      const solutionHash = publicOutputs.solutionHash;
+
+      expect(solutionHash).not.toEqual(computedHash);
+
+      expect(BigInt(round)).toEqual(
+        publicOutputs.turnCount.sub(1).div(2).toBigInt()
+      );
+    });
```

Comment:
You can increase the timeout as shown below to help the tests pass:
```suggestion
    }, 2000000);
```

However, the newly added tests are taking quite a long time to run, as shown in the logs:

```sh
 Should generate StepProgramProof for given parameters
      ✓ Should generate proofs and we shall obtain of codeMaster win. (859897 ms)
      ✓ Should generate a game where codeBreaker wins (417426 ms)
      ✓ Should generate a proof that is not solved yet. (420486 ms)

Test Suites: 1 passed, 1 total
Tests:       37 skipped, 3 passed, 40 total
Snapshots:   0 total
Time:        1713.142 s
```

Given these long execution times, we might need to consider reducing the number of rounds or optimizing the test scenarios for better performance :thinking:

Reply:
With using `proofsEnabled:false` as
```ts
await StepProgram.compile({
proofsEnabled: false
})
```
all test took 8 seconds to finish
```bash
Test Suites: 1 passed, 1 total
Tests:       40 passed, 40 total
Snapshots:   0 total
Time:        8.688 s
Ran all test suites.
```
but the StepProgram must be previously cached in order to use it. Other than that it might be good to reduce the number of rounds.

Reply:
Oh nice! I thought `proofsEnabled: false` doesn't work for recursive proofs!